### DISCUSSION TITLE
Sharing AnalyzerConfig between instances of AbstractUserAgentAnalyzer

### DISCUSSION THREAD

**Discussion by christhill** [nielsbasjes/yauaa#1223] [Ideas]
Creating a new instance of a `UserAgentAnalyzer` creates a new instance of an `AnalyzerConfig`.  The creation of a new `AnalyzerConfig` reads a large number of YAML files and takes up approximately 20MG of the stack.  This slows down the creation of a `UserAgentAnalyzer` and increases the memory used if you have more than one instance of a `UserAgentAnalyzer`.  As far as I can tell, the `AnalyserConfig` is static and does not change.

1. Is it safe to share the AnalyserConfig between UserAgentAnalysers?
2. Can the Builder be modified to allow the reuse of the AnalyserConfig between instances of UserAgentAnalyzer?

**Comment by nielsbasjes** [comment]
My first question to you is why this ~20MiB of memory usage is an issue to you? 

The runtime is a few hundred MiB per instance (not sharable) and to me this 20MiB seems like a very small part of this. Also you should not have more than a handful of instances per JVM and having 5 instances would mean ~100MiB too much which in the systems where I run it is a tiny value.

Also I have done benchmarks and most of the time is not spent in loading these files. Most of the time is spent in creating the (not sharable) data structures. You can try to speed things up by using Kryo and serialize the instance and then deserialize it as an identical clone. This should give you a speedup.

**Reply by christhill** [reply]
Hi Niels,

The JVM we run YAUAA in uses about 1.5 GiB of heap.  A shared config would reduce the memory usage for each additional YAUAA instance from 128MiB to 100MiB. While the 20 MiB of config is a manageable amount of memory, it seems an easy win to reuse the configuration object and free that memory. (NOTE: We don't use the cache built into YAUAA.  This allows us to share a cache across YAUAA instances and only cache a subset of the `ImmutableUserAgent` properties to save space)

We run two instances of YAUAA and a large cache, which mostly works well.  However, we have times when parsing the user agent is a bottleneck, mostly during instance startup when the cache is empty, and we need to parse the user agent for most requests.  It is a balance between the memory usage of more instances and the maximum user-agent parse time we can tolerate.  For example, sharing the config would allow us to have 6 YAUAA instances for the memory footprint of 5 YAUAA instances without config sharing.

Some rough testing showed that the built time for each YAUAA instance was about 250ms.  If we reuse the config, this can be reduced to 3ms.  When building the YAUAA instance, I set delayedInit = true, so the build time did not include the matcher initialisation (about another 250ms).

The class I used for this testing was
```
    public static class CustomUserAgentAnalyzer extends AbstractUserAgentAnalyzer {
        public CustomUserAgentAnalyzer(AnalyzerConfig analyzerConfig) {
            configure(analyzerConfig, false, true);
            wasBuilt = true;
            setClientHintsCacheSize(0);
            setCacheSize(0);
        }
    }
```

**Reply by nielsbasjes** [reply]
I've been thinking about your suggestion and I'm currently trying something out.
https://github.com/nielsbasjes/yauaa/tree/CloneWithSharedConfig

**Reply by nielsbasjes** [reply]
If I create a new instance each time I see this:

```
After 1: DELTA = 35173568 bytes
After 2: DELTA = 33984592 bytes
After 3: DELTA = 34117536 bytes
After 4: DELTA = 34115256 bytes
```

if I clone them sharing the config

```
After 1: DELTA = 34050032 bytes
After 2: DELTA =  3141296 bytes
After 3: DELTA =  3141296 bytes
After 4: DELTA =  3141296 bytes
```

I'm still checking if everything is still working correctly.

**Reply by christhill** [reply]
That looks great.  Looking forward to using it.

**Reply by nielsbasjes** [reply]
Released https://github.com/nielsbasjes/yauaa/releases/tag/v7.24.0
